docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,152 @@
+    "pre": {
+      "failurePolicy": "Abort",
+      "execNewPod": {
+        "containerName": "test-container",
+        "command": [
+          "/opt/app-root/bin/verify-deployment"
+        ],
+      }
+    }
+  }
+}
+```
+
+Based on the result of the `verify-deployment` command, that runs in the application
+image, the command will promote the image by tagging it as `production-ready` (eg. using
+the `oc tag` command). The command might also notify me about the verification failure.
+
+The "frontend-production" DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the "prod-ready" ImageStreamTag. When a new Docker image with this tag is available,
+it can automatically rolls out a new production deployment. In case you don't want to roll out
+the "frontend-production" automatically, you can avoid specifying the ImageChangeTrigger and
+roll out the new deployment by hand.
+
+### Manual promotion using external Docker registry
+
+As a developer of a Rails application, when I update the source code and OpenShift
+build my image I want to promote the image by passing it throught QA and devops team.
+
+#### Example
+
+* The image is passed to a remote QA team by pushing the image to a shared Docker registry
+* Remote QA team pull the image and drive the testing and verification
+* Remote QA team push the verified image back to the shared Docker registry, tagged as "verified" and with additional sign-off LABEL(s).
+* Remote devops team pull the image from shared Docker registry and make it available in "stage" cluster.
+* Remote devops team promotes the image from "stage" to "production"
+
+### Manual promotion of builder image using external Docker registry
+
+As a platform operator, I want to be able to promote the S2I builder image(s) when a new
+version is available in Red Hat Docker registry. I want to perform additional verifications
+(iow. deploy to "stage" first) before rolling a new image out to production which causes rebuild
+of thousand applications.
+
+#### Example
+
+* OpenShift engineering push the `openshift/ruby-22-rhel7` image to internal CI registry.
+* OpenShift QA team pull the image from the internal CI registry and run tests.
+* OpenShift QA team sign-off the Docker image when tests pass and push it to "integration" registry.
+* OpenShift operations team pulls the image from "integration" registry and promote it to "stage" environment.
+* OpenShift operations team manually promote the image to production.
+
+### Semi-automatic promotion and testing of builder images
+
+As a OpenShift S2I image developer, I want to have the image rebuilt, tested and pushed to
+the Docker Hub or to internal registry and tag it as "qa-ready" and notify the QA team about
+availability of a new Docker Image.
+
+#### Example of pull request tester
+
+* OpenShift engineering propose a change to `openshift/ruby-22-rhel7` image.
+* CI starts OpenShift server and execute the extended tests cases related to this image.
```

Comment:
@bparees yeah, I think we will need to have some OpenShift server up and running that CI can connect to. Maybe as part of our Jenkins setup. @danmcp ?